### ISSUE TITLE
"..." is not iterable when calling a conditional return function

### ISSUE THREAD

**Issue by PrateekArya** [microsoft/pylance-release#54]
## Environment data

-   Language Server version: 2020.6.1
-   OS and version: 10 and 1909
-   Python version (& distribution if applicable, e.g. Anaconda): Anaconda 3, python 3.6.5 64-bit

## Expected behaviour

No error

## Actual behaviour

![image](https://user-images.githubusercontent.com/1255663/86523006-5e85f380-be66-11ea-85d0-b43968646a65.png)

## Logs

```
[FG] parsing: d:\...\main4NewAdm.py (234ms)
[FG] binding: d:\...\main4NewAdm.py (10ms)
[FG] parsing: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\display.py (16ms)
[FG] binding: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\display.py ...
[FG]   parsing: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\display.py (202ms)
[FG]   binding: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\core\display.py (9ms)
[FG]   parsing: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\lib\display.py (75ms)
[FG]   binding: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\lib\display.py (5ms)
[FG] binding: C:\Users\...\AppData\Local\Continuum\anaconda3\lib\site-packages\IPython\display.py (291ms)
IntelliCode: current invocation did not produce any meaningful tokens.
Time taken to get recommendations: 116 ms, Memory increase: 490 KB.
[BG] analyzing: d:\...\main4NewAdm.py ...
[BG]   parsing: d:\...\main4NewAdm.py (221ms)
[BG]   binding: d:\...\main4NewAdm.py (24ms)
[BG]   checking: d:\...\main4NewAdm.py (1776ms)
[BG] analyzing: d:\...\main4NewAdm.py (2021ms)
[FG] parsing: d:\...\main4NewAdm.py (37ms)
[FG] binding: d:\...\main4NewAdm.py (5ms)
[BG] analyzing: d:\...\main4NewAdm.py ...
[BG]   parsing: d:\...\main4NewAdm.py (71ms)
[BG]   binding: d:\...\main4NewAdm.py (3ms)
[BG]   checking: d:\...\main4NewAdm.py (52552ms)
[BG] analyzing: d:\...\main4NewAdm.py (52626ms)
```

## Code Snippet / Additional information

![image](https://user-images.githubusercontent.com/1255663/86523005-5332c800-be66-11ea-9bf5-532c5fc75952.png)

**Comment by erictraut** [comment]
Thanks for reporting the issue, but there isn't enough information here for me to investigate further. I would need a self-contained code snippet or instructions for how to repro.

Based on the screen shot above, it looks like there is a code path whereby importedStuff can contain None. This will result in an exception if you attempt to assign it a tuple. So without any additional details, it appears that this is a bug in your code that Pyright has helped to identify. I recommend looking at the function `dPP.importStuff()` to determine whether there is a code path by which it returns `None`. If so, your code should be prepared to deal with that return value.

**Comment by PrateekArya** [comment]
I understand your argument about my code possibly returning `None` which is actually possible (refer the importStuff function below). In that case my code will throw an exception which is what I want it to do. :) But I think if Pylance/Pyright wants to point it out then it should categories it as a `warning` or `information` as this is not really an error!

`#Function to import (exported using above function) stuff
def importStuff(fileName='ExportedObjs', ext='', path=''):

    #Adding path splitter if not supplied
    if not len(path) == 0 and not path[-2:] == '\\':
        path += '\\'

    #Checking extension
    if ext == '':
        dotIdx = fileName[-5:].find('.')
        if dotIdx == -1:
            ext = '.pkl'    #Setting default extension
        else:  
            #Extracting extension from file name
            ext = fileName[-5:][dotIdx:]
            fileName = fileName[:-5+dotIdx]

    #Reading pickle file
    if ext == '.pkl':
        #Reading file
        fRead = open(path + fileName + ext, 'rb')

        importList = []
        #Loading objects one by one
        try:
            while True:
                importList.append(pkl.load(fRead))
        except EOFError:
            fRead.close()
            return importList
    elif ext == '.xls' or ext == '.xlsx':
        #Reading excel
        fileObj = pd.ExcelFile(path + fileName + ext)

        data = dict({})
        #Reading all sheets in a Dict
        for sheetName in fileObj.sheet_names:
            data[sheetName] = pd.read_excel(fileObj, sheetName, index_col=False)
        
        if len(data.keys()) == 1:
            return data[sheetName]
        else:
            return data
    elif ext == '.csv':
        #Reading csv
        return pd.read_csv(path + fileName + ext)
`

**Comment by erictraut** [comment]
This particular diagnostic ("x is not iterable") should be associated with the diagnostic rule "reportGeneralTypeIssues". Currently, it's not. I'll make that change. Because of this, it is reported even if typeCheckingMode is set to "off". It should be reported only if typeCheckingMode is set to "basic" or "strict".

**Comment by PrateekArya** [comment]
Please feel free to close this then. I'll be on vacations for next two weeks. 🙂

**Comment by jakebailey** [comment]
This issue has been fixed in version 2020.7.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/master/CHANGELOG.md#202070-9-july-2020
